src/antsibull/versions.py

```diff
@@ -0,0 +1,144 @@
+            if self.upper_operator is not None:
+                raise ValueError(
+                    f"Multiple upper version limits specified for {self.collection_name}:"
+                    f" {self.spec}"
+                )
+            self.upper_operator = clause.operator
+            self.upper_version = clause.target
+            # Omit the upper bound as we're replacing it
+            return
+
+        if clause.operator == ">=":
+            # Save the lower bound so we can write out a new compatible version
+            if self.min_version is not None:
+                raise ValueError(
+                    f"Multiple minimum versions specified for {self.collection_name}: {self.spec}"
+                )
+            self.min_version = clause.target
+
+        if clause.operator == ">":
+            raise ValueError(
+                f"Strict lower bound specified for {self.collection_name}: {self.spec}"
+            )
+
+        if clause.operator == "==":
+            self.pinned = True
+
+        self.potential_clauses.append(clause)
+
+    def compute_new_spec(self) -> str:
+        """
+        Compute a new version range specification that only allows newer patch releases that also
+        match the original range specification.
+        """
+        if self.pinned:
+            if len(self.clauses) > 1:
+                raise ValueError(
+                    f"Pin combined with other clauses for {self.collection_name}: {self.spec}"
+                )
+            return self.spec
+
+        upper_operator = self.upper_operator
+        upper_version = self.upper_version
+        if upper_operator is None or upper_version is None:
+            raise ValueError(
+                f"No upper version limit specified for {self.collection_name}: {self.spec}"
+            )
+
+        min_version = self.min_version
+        if min_version is None:
+            raise ValueError(
+                f"No minimum version specified for {self.collection_name}: {self.spec}"
+            )
+
+        if min_version.next_minor() <= upper_version:
+            upper_operator = "<"
+            upper_version = min_version.next_minor()
+
+        new_clauses = sorted(
+            str(clause)
+            for clause in self.potential_clauses
+            if clause.target < upper_version
+        )
+        new_clauses.append(f"{upper_operator}{upper_version}")
+        return ",".join(new_clauses)
+
+
+def feature_freeze_version(spec: str, collection_name: str) -> str:
+    """
+    Make semantic version range specification valid for feature freeze.
+    """
+    return _FeatureFreezeVersion(spec, collection_name).compute_new_spec()
+
+
+def load_constraints_if_exists(filename: str) -> dict[str, SemVerSpec]:
+    """
+    Load a constraints file, if it exists.
+    """
+    result: dict[str, SemVerSpec] = {}
+    if not os.path.exists(filename):
+        return result
+    for line in parse_pieces_file(filename):
```

Comment:
```suggestion
    for line in parse_pieces_file(os.fspath(filename)):
```

then. I'm also fine with leaving this for a separate PR.